### PR TITLE
Introduce group_replica partial response strategy

### PR DESCRIPTION
**Pull request by hczhu-db**
This is to implement the idea in https://docs.google.com/document/d/1eGoUFNGq2pOhmQH_svBkUZ45ig8bVOZH6xi9wkNRN_E/edit#heading=h.axx9hx8o4mz5

This strategy takes effect only when the partial response tolerance flag is off. Therefore, by default, this strategy is off.

Tested the image in `dev-aws-eu-west-1` without turning on the new strategy.

# How does the new strategy work?
1. A repeatable command line flag `endpoint` specifies multiple DNS names, like `dnssrv+db-rep0:9092`, `dnssrv+db-rep1:9092`, and `dnssrv+range-store:9092`
```
endpoints := extkingpin.Addrs(cmd.Flag("endpoint", "Addresses of statically configured Thanos API servers (repeatable). The scheme may be prefixed with 'dns+' or 'dnssrv+' to detect Thanos API servers through respective DNS lookups.").
		PlaceHolder("<endpoint>"))
```
2. Querier resolves each DNS name and gets a set of endpoints. Each endpoint set has a replica key and a group key parsed from the DNS name according to a pre-defined format.
```
Group key: db
Replica key: db-rep0
Endpoints: IP0, IP1, ...
---
Group key: db
Replica key: db-rep1
Endpoints: IP0, IP1, ...
---
Group key: range-store
Replica key: rang-store
Endpoints: IP0, IP1, ...
```
3. Querier fans out a query to all endpoints, even if one is determined unhealthy (different from other strategies). Querier counts how many failures from each group and replica and determines if the query result is actually complete.

### CODE REVIEW

File: cmd/thanos/query.go [hunk 1]
```diff
@@ -179,6 +180,8 @@ func registerQuery(app *extkingpin.App) {
 		Default(string(dns.MiekgdnsResolverType)).Hidden().String()
 
 	unhealthyStoreTimeout := extkingpin.ModelDuration(cmd.Flag("store.unhealthy-timeout", "Timeout before an unhealthy store is cleaned from the store UI page.").Default("5m"))
+	ignoreStoreErrors := cmd.Flag("store.ignore-errors", "Ignore errors from store endpoints. This is for qeury result completeness.").
```
Comment (jnyi) [inline cmd/thanos/query.go]: what's the reason to ignore errors from store? not clear on the rational from the comment

File: pkg/cacheutil/memcached_client_test.go [hunk 2]
```diff
@@ -750,7 +750,7 @@ func TestMemcachedClient_SetAsync_CircuitBreaker(t *testing.T) {
 				// Trigger the state transaction.
 				time.Sleep(time.Millisecond)
 				testutil.Ok(t, client.SetAsync(strconv.Itoa(testdata.setErrors), []byte("value"), time.Second))
-				testutil.Equals(t, gobreaker.StateOpen, cbimpl.State(), "state should be open")
+				// testutil.Equals(t, gobreaker.StateOpen, cbimpl.State(), "state should be open")
```
Comment (jnyi) [inline pkg/cacheutil/memcached_client_test.go]: any reason to comment this test?

File: cmd/thanos/query.go (hunk 1 above)
Comment (jnyi) [reply cmd/thanos/query.go]: oh, does here store not refer to `pantheon-store`, but any components enable storeAPI? If this is a feature flag control, then ignore my comment

File: pkg/query/endpointset.go [hunk 3]
```diff
@@ -44,12 +46,25 @@ type queryConnMetricLabel string
...
 	ExternalLabels queryConnMetricLabel = "external_labels"
 	StoreType      queryConnMetricLabel = "store_type"
+	GroupKey       queryConnMetricLabel = "group_key"
+	ReplicaKey     queryConnMetricLabel = "replica_key"
 )
 
+var gReplicaKeySuffixRegex *regexp.Regexp
+
+func init() {
+	gReplicaKeySuffixRegex = regexp.MustCompile(`-rep[0-9]$`)
```
Comment (jnyi) [inline pkg/query/endpointset.go:56]: aha, this seem very specific to our setup :)

**Review by jnyi — APPROVED** [review]
LGTM, nice, just a few small comments, do we have any tests to see how this behaves? Also, please make sure e2e tests pass to avoid regressions

File: cmd/thanos/query.go (hunk 1 above)
Comment (hczhu-db) [reply cmd/thanos/query.go]: Correct. I found in Thanos code base that `store` refers to both Store and DB which both provide time series streaming API.
Yes, it's a feature flag. I'll update the flag summary.

File: pkg/query/endpointset.go (hunk 3 above)
Comment (hczhu-db) [reply pkg/query/endpointset.go:56]: yes, it's a hacky solution. 
image

**Comment by hczhu-db** [comment]
> LGTM, nice, just a few small comments, do we have any tests to see how this behaves? Also, please make sure e2e tests pass to avoid regressions

some end-to-end tests are flaky on db_main.

File: pkg/cacheutil/memcached_client_test.go (hunk 2 above)
Comment (hczhu-db) [reply pkg/cacheutil/memcached_client_test.go]: I reverted this change.

**Pull request merged**